### PR TITLE
Add embedded tar script packager

### PR DESCRIPTION
**Pull request by oyarzun**
Adds an embedded tar script packager.

- Installer run as normal user it defaults to `~/.local/appdir`
- Installer run with sudo it defaults to `/usr/local/appdir`
- Install desktop file

### CODE REVIEW

File: src/main/java/org/apache/netbeans/nbpackage/NBPackage.java [hunk 1]
```diff
@@ -132,7 +133,8 @@ public final class NBPackage {
             new RpmPackager(),
             new InnoSetupPackager(),
             new PkgPackager(),
-            new ZipPackager()
+            new ZipPackager(),
+            new TarScriptPackager()
```
Comment (neilcsmith-net) [inline src/main/java/org/apache/netbeans/nbpackage/NBPackage.java:137]: Keep zip last - move tar below rpm to keep Linux options together.

File: src/main/resources/org/apache/netbeans/nbpackage/shell/shell.script.template [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+if ! [ $(id -u) = 0 ]; then
+   DEFAULT_INSTALL_PREFIX=$HOME/.local
+   DESKTOP_FILE_INSTALL_DIR=$HOME/.local/share/applications
+else
+   DEFAULT_INSTALL_PREFIX=/usr/local
+   DESKTOP_FILE_INSTALL_DIR=/usr/local/share/applications
+fi
+
+read -p "Enter ${APP_NAME} Install Prefix [${{DEFAULT_INSTALL_PREFIX}}]/${{APP_DIR}}: " INSTALL_PREFIX
```
Comment (neilcsmith-net) [inline src/main/resources/org/apache/netbeans/nbpackage/shell/shell.script.template:13]: Move to `${DEFAULT_INSTALL_PREFIX}` pattern.

Also, mixing up `${{APP_DIR}}` and `${APP_DIR}`.

**Review by neilcsmith-net — CHANGES_REQUESTED** [review]
Thanks, interesting idea.  Sorry for the delay.  I've not got a lot of time for NetBeans things at the moment, but had a glance through.  Few concerns, but nothing major.

We were discussing getting a first non-beta release out, which involves removing a few deprecated options first.  Not sure if this should merge before or after that release.

I was wondering whether we could reduce duplication of some of these options by having some shared Linux options.  That might require some thought given the current structure, and might not be worth the trouble.  Depends if we might expect more Linux package types in future.

File: src/main/java/org/apache/netbeans/nbpackage/shell/TarScriptPackager.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    static final ResourceBundle MESSAGES
+            = ResourceBundle.getBundle(TarScriptPackager.class.getPackageName() + ".Messages");
+
+    /**
+     * Path to png icon (48x48) as required by AppDir / XDG specification.
+     * Defaults to Apache NetBeans icon.
+     */
+    static final Option<Path> ICON_PATH
+            = Option.ofPath("package.shell.icon",
```
[the remark this answers is not in the document]
Comment (oyarzun) [reply src/main/java/org/apache/netbeans/nbpackage/shell/TarScriptPackager.java:43]: Sounds good, I was having trouble deciding what to call the package.

File: src/main/java/org/apache/netbeans/nbpackage/StringUtils.java [hunk 4]
```diff
@@ -30,11 +30,31 @@
...
         // static utilities
     }
 
+  /**
+    * Unescapes the bash variables in the script template
+    *
+    * @param template script template
+    * @return script with unescaped variables
+    */
+    public static String unescapeScriptVars(String script)
```
[the remark this answers is not in the document]
Comment (neilcsmith-net) [reply src/main/java/org/apache/netbeans/nbpackage/StringUtils.java:45]: Sure, I understand what you're trying to do, but that should happen anyway without this extra code as long as the token is not present during packaging - eg. see the test at https://github.com/apache/netbeans-nbpackage/blob/51e9d3db6a3438f85f141817d4cd9a97de9b3cc4/src/test/java/org/apache/netbeans/nbpackage/ExecutionContextTest.java#L77-L79

[the remark this answers is not in the document]
Comment (neilcsmith-net) [reply src/main/java/org/apache/netbeans/nbpackage/StringUtils.java:45]: Yes, there's `StringUtils::replaceTokens` where you want to leave any unmatched tokens as they are, and `StringUtils::replaceTokensOrFail` for where all tokens must be pre-defined.  It was done like this to allow shell scripts and other contexts where the same `${...}` is needed.

Ideally, think about the naming in the script templates to make it clear which are replaced at build time though if you need both.  Maybe the additional tokens like `APP_DIR` should be lower case in this one?

**Comment by oyarzun** [comment]
@neilcsmith-net Happy New Year! 

Any more thoughts on this PR? Does it need more changes? It it worthwhile to be merged?

**Comment by neilcsmith-net** [comment]
Thanks @ebarboni but I don't agree with you that the version can be decided later.  If we want the next release to be 1.0.0 then we need to do that _before_ this is merged.

**Comment by neilcsmith-net** [comment]
I'll do some further testing soon, and see if we can merge now that 1.0 is released.  Would still appreciate another set of eyes on this - maybe @mbien or @homberghp have some thoughts?

**Review by mbien — COMMENTED** [review]
Hi! Sorry for the delays.

1) Code looks great. 

2) Tried a minimal setup to bundle NetBeans Deluxe Edition. For that I only set the name in the properties file, I think according to the comments in the config thats all what is needed as minimal setup.

bundle creation worked without warning, but on install on arch/gnome I got:

```bash
$ ./NB_Deluxe-1.0.noarch.sh 
Enter NB Deluxe Install Prefix [/home/virtual1/.local]/netbeans: mega
Installing NB Deluxe in mega/netbeans
/home/virtual1/.local/share/applications/netbeans.desktop: error: value "mega/netbeans/launcher/netbeans.svg" for key "Icon" in group "Desktop Entry" looks like a relative path, instead of being an absolute path to an icon or an icon name
Error on file "mega/netbeans/launcher/netbeans.desktop": Failed to validate the created desktop file
```
It appears that it doesn't like the rel path of the fallback icon?

3) the "Supported packagers" section of the readme would need adjusting i think and also list the new type.

**Comment by neilcsmith-net** [comment]
Thanks for looking @mbien The desktop file spec says icons must be absolute paths or the name to look up.  Given that this script already handles desktop file installation into `/usr/local/share` or `$HOME/.local/share` it might be good to copy the png and svg icons to the correct places instead?

Given no review from Pieter still, cc'ing @geertjanw instead.

**Comment by oyarzun** [comment]
@neilcsmith-net @mbien  Thanks for reviewing the PR. I've rebased the code and addressed @mbien comments:

1. Use `realpath` to resolve the absolute install path used in the desktop file (e.g. the icon path)
2. Update the readme with the new package type

**Review by mbien — APPROVED** [review]
works, thanks! There would be still the question of how to uninstall it again, but I think this should probably be fine as initial implementation.

**Comment by oyarzun** [comment]
> works, thanks! There would be still the question of how to uninstall it again, but I think this should probably be fine as initial implementation.

I suppose the install script could create an uninstall script that deletes the directory and the desktop file.

**Comment by mbien** [comment]
right I had the same thought. What I also haven't tested is what happens when the folder is non-empty, the script will merge the content without warning I believe. Script should probably also fail when someone passes arguments to it, e.g "I am an interactive installer, no arguments expected.". But those things could be followups I suppose, no need to delay this PR even further.

**Comment by mbien** [comment]
features would also have the potential to build on top of each other. E.g if the folder is non empty it could check for the uninstaller and run it first. If there is none, it would refuse installing. That way it could upgrade in place.

**Comment by neilcsmith-net** [comment]
> That way it could upgrade in place.

That's a point - all existing packagers are designed to allow upgrade in place.

> no need to delay this PR even further.

I would like to see some input from FoAN people here before this is merged.
